### PR TITLE
for issue #326, added ShouldContainMatchingCount

### PR DESCRIPTION
**Pull request by mgroves**
I'm sure the unit tests and/or test messages could be improved, but I think commit will give you the Should extension that you want. This is for issue #326

### CODE REVIEW

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs [hunk 1]
```diff
@@ -47,6 +47,15 @@ public static void ShouldContain<T>(this IEnumerable<T> actual, [InstantHandle]
         {
             ShouldContain(actual, elementPredicate, () => null);
         }
+        public static void ShouldContainMatchingCount<T>(this IEnumerable<T> actual, [InstantHandle] Expression<Func<T, bool>> elementPredicate, int expectedCount)
```
Comment (JakeGinnivan) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: I think this should be an overload of `ShouldContain` for discovery?

File: src/Shouldly.Tests.Shared/ShouldContain/CollectionPredicateWithCountScenario.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        [Fact]
+        public void CollectionWithACountOtherThanTheSpecifiedCountOfMatchingPredicatesFails()
+        {
+            var collection = new[] { "a", "b", "c", "c" };
+            Verify.ShouldFail(() =>
+            collection.ShouldContainMatchingCount(x => x == "c", 5),
+
+errorWithSource:
+@"[""a"", ""b"", ""c"", ""c""]
```
Comment (JakeGinnivan) [inline src/Shouldly.Tests.Shared/ShouldContain/CollectionPredicateWithCountScenario.cs]: With source it should be `collection` at the start?

**Comment by JakeGinnivan** [comment]
Make sure you are running the tests under debug build, not release

**Comment by JakeGinnivan** [comment]
Thanks for the PR :)

**Comment by mgroves** [comment]
My bad; some sort of local snafu in saving a file. New commit on the way.

**Comment by mgroves** [comment]
Well I sure am making a mess. Now I'm failing convention tests. I'll work on it; sorry for the noise :)

**Comment by JakeGinnivan** [comment]
Not a problem, am working on GitVersion atm so am around if you have any questions or need a hand

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs (hunk 1 above)
Comment (mgroves) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: Whoops, didn't see this comment. I'll rename; is this also causing a test failure?

**Comment by JakeGinnivan** [comment]
There is a test which approves the public API. You just have to run that test manually and it will pop-up a difftool which you merge your API changes in.

The reason for this is so it makes it super easy to review public API changes (so we don't accidentally make breaking changes)

**Comment by mgroves** [comment]
Ah, interesting, I've not seen that before.

**Comment by JakeGinnivan** [comment]
Latest Shouldly feature includes the feature - http://shouldly.readthedocs.org/en/latest/assertions/shouldMatchApproved.html :)

**Comment by mgroves** [comment]
There we go. Thanks for helping me; I hope this allows you to close out an issue, and maybe @AmadeusW will get some use out of it :)

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs [hunk 3]
```diff
@@ -48,6 +48,26 @@ public static void ShouldContain<T>(this IEnumerable<T> actual, [InstantHandle]
...
+        {
+            var condition = elementPredicate.Compile();
+            var actualCount = actual.Count(condition);
+            if (actualCount != expectedCount)
+            {
+                throw new ShouldAssertException(new ExpectedActualShouldlyMessage(expectedCount, actual, () => null).ToString());
+            }
+        }
+
+        public static void ShouldContain<T>(this IEnumerable<T> actual, [InstantHandle] Expression<Func<T, bool>> elementPredicate, int expectedCount, string customMessage)
```
Comment (JakeGinnivan) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: I think the `Func<string> customMessage` overload is missing.

Also you can make each of these methods call each other.

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs [hunk 4]
```diff
@@ -48,6 +48,26 @@ public static void ShouldContain<T>(this IEnumerable<T> actual, [InstantHandle]
...
+        {
+            var condition = elementPredicate.Compile();
+            var actualCount = actual.Count(condition);
+            if (actualCount != expectedCount)
```
Comment (JakeGinnivan) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: I think the logic is wrong for this, I believe the overload should be counting the number of items matched by the predicate.

as an example

var arr = new[] { 1, 2, 3, 4 };
arr.ShouldContain(i => i %2, 2); // Should have exactly 2 items which matches predicate

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs (hunk 3 above)
Comment (mgroves) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: Yep, that makes sense. Added that in another commit.

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs (hunk 4 above)
Comment (mgroves) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: I think the logic is correct. I added your example in the unit test (although I assumed that you meant:

```
arr.ShouldContain(i => (i %2) == 0, 2);
```

i.e. a predicate to get even numbers).

**Comment by JakeGinnivan** [comment]
There are a few more things (which will also make the error message awesome).

When creating the exception, do this:

```
// Create the type in ShouldlyMessage.cs
internal class ShouldContainWithCountShouldlyMessage : ShouldlyMessage
{
    public ShouldContainWithCountShouldlyMessage(object expected, object actual, int matchCount, [InstantHandle] Func<string> customMessage, [CallerMemberName] string shouldlyMethod = null)
    {
        ShouldlyAssertionContext = new ShouldlyAssertionContext(shouldlyMethod, expected, actual)
        {
            HasRelevantActual = true,
            MatchCount = matchCount
        };
        if (customMessage != null) ShouldlyAssertionContext.CustomMessage = customMessage();
    }
}

// Throw this instead
throw new ShouldAssertException(new ShouldContainWithCountShouldlyMessage(elementPredicate.Body, actual, expectedCount, customMessage).ToString());

// Finally, update the ShouldContainPredicateMessageGenerator to use MatchCount if set in the error message
```

Also, would you be able to update the docs for this guy :P I am not sure if ShouldContain has been pulled across yet, but if you are up for migrating those docs then adding this then that wraps this up entirely!

File: src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs (hunk 4 above)
Comment (JakeGinnivan) [inline src/Shouldly.Shared/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs]: You are right, the error message confused me. Have put another comment how to fix the error message :)

**Comment by mgroves** [comment]
Geez I thought this was going to be easy :)

Just committed the improved error messaging. I'll take a look at docs

**Comment by JakeGinnivan** [comment]
Yeah, sorry about that. I am heading out for a few hours. Will get back to this once back

**Comment by JakeGinnivan** [comment]
But changes are looking good!

**Comment by mgroves** [comment]
That latest build fail seems to be choking on "x = c" vs "x == c". I don't know it would be generating single vs double equals? Seems to be passing locally just fine. Any ideas?

It looks like docs are using "Sphinx" or something? I've not heard of it, but I'm willing to give it a shot if I have time, but it may be a good idea to open up a separate issue for that in case I can't get to it right away?

**Comment by JakeGinnivan** [comment]
@mgroves I finally had a chance to look into this

The reason the net35 tests are failing is that net35 doesn't have expression visitors, so we can't format the expression tree nicely. I am skipping these tests for net35.

**Comment by JakeGinnivan** [comment]
And yeah, the docs are in sphinx. It is kinda like markdown just a lot more powerful.

It allows us to pull in external content, so when we update our error messages in shouldly out docs automatically get the changes :) You should be able to follow the patterns we have pretty easily to add docs.

**Comment by JakeGinnivan** [comment]
Thanks for all your work, it will be in the 2.7.0 release (beta 2 will be out within the hour)

**Pull request merged**